### PR TITLE
Allow "readonly" and "readonly no download" to edit widgets

### LINKED ISSUE #831: Allow readonly users to edit saved chart titles.
We want to fix an error on the widgets endpoint that forbids readonly users from editing saved chart metadata, since they are already able to save charts they should be able to edit the saved chart metadata.

### PR DESCRIPTION
**Pull request by urbanslug**
closes #831 
closes https://github.com/onaio/zebra/issues/4685

### CODE REVIEW

File: onadata/apps/api/permissions.py [hunk 1]
```diff
@@ -277,6 +277,9 @@ def has_object_permission(self, request, view, obj):
                 isinstance(obj.content_object, DataView)):
             return False
 
+        if 'key' in request.query_params or view.action == 'partial_update':
```
Comment (pld) [inline onadata/apps/api/permissions.py]: how does this work? is it still validating the user has permissions on the form this widget is attached to?

Comment (urbanslug) [inline onadata/apps/api/permissions.py]: No it is not. I am looking for a way to find the user's permission in the parent Xform and restrict that in the creation and changing of the widget.

**Comment by denniswambua** [comment]
Updated to check xform perms

File: onadata/apps/api/tests/viewsets/test_widget_viewset.py [hunk 2]
```diff
@@ -210,6 +211,81 @@ def test_list_widgets(self):
...
+                              self.xform.pk,
+            'description': "Test widget",
+            'aggregation': "Sum",
+            'widget_type': "charts",
+            'view_type': "horizontal-bar",
+            'column': "age",
+            'group_by': ''
+        }
+
+        # to do: test random user with auth
```
Comment (pld) [inline onadata/apps/api/tests/viewsets/test_widget_viewset.py]: can we do this now?

Comment (denniswambua) [inline onadata/apps/api/tests/viewsets/test_widget_viewset.py]: done

**Comment by denniswambua** [comment]
yes, i have added a test case for user without perms.

**Pull request merged**